### DISCUSSION TITLE
Fill route with shaping points with adjustable distance

### DISCUSSION THREAD

**Discussion by devemux86** [devemux86/cruiser#499] [Q&A] [answered]
Tool for adding extra shaping points to routes.

Useful in navigation to [reroute back to original route as soon as possible](https://github.com/devemux86/cruiser/discussions/497) when the route has few waypoints.

- Keeps original via and shaping points

- The distance can be selected

- Named with distance to destination

- Undo can revert changes

- Works **offline** without route calculation

Before|After
-|-
<img src="https://github.com/devemux86/cruiser/assets/3484020/7c7c32f3-f941-4a5a-9bc5-7db54f5f55a2"/>|<img src="https://github.com/devemux86/cruiser/assets/3484020/befdc0b8-5d94-4d7a-a36a-4f1a85fe6790"/>

**Comment by devemux86** [comment]
What values do you think the distance slider can have?
- Default: 10 km
- Minimum: 1 km
- Maximum: 50 km

**Reply by devemux86** [reply]
Also should the added shaping points have a name?
e.g. the trip distance `58.7 km`

**Reply by 0709wiwiwi** [reply]
I don't really know. If applied since start I had initially thought 010 020 etc if per 10km but after a recalculation this is of course no longer correct.  Quick idea. Add from the end to the start ? Still xxx km to go.    _Countdown to zero_

**Reply by devemux86** [reply]
@0709wiwiwi you are right.

Automatic names can often complicate things. :slightly_smiling_face:

**Reply by WolfgangZ42** [reply]
> Add from the end to the start ? Still xxx km to go.

That's exactly what I do with gpxshaping.de right now. Would appreciate that and could then use Cruiser instead. 
Example: [0599_München SO.zip](https://github.com/devemux86/cruiser/files/15027381/0599_Munchen.SO.zip)

**Reply by bios4** [reply]
Nice feature!

Again a suggestion regarding wording: I think "Add shaping points" could easily be misinterpreted with the existing "Set shaping point" feature where one can freely add shaping points. Therefore I would prefer to read something like "Auto-fill with shaping points" or something similar.

What do you think?

> What values do you think the distance slider can have?

I think the options to auto-fill with shaping pints "every X kilometer(s)" with X = {1, 2, 5, 10, 25, 50} could be quite useful.

> Also should the added shaping points have a name?

Either "X km from start" or "X km to destination" would be nice, maybe as an either-or setting?
Otherwise just number them...

And maybe it is possible to give them a different color in comparison to the manually set shaping points!??

**Reply by 0709wiwiwi** [reply]
To think about fresh idea in the morning ;-)

> Therefore I would prefer to read something like "Auto-fill with shaping points" or something similar.

Add (countdown named) distance markers ? 

> And maybe it is possible to give them a different colour in comparison to the manually set shaping points!???

If you take that step, you could possibly show the temporary autofill distance marker shaping points as discrete mini black dots.
They are thus not very overwhelmingly distracting on the map but are still clearly visible in the waypoint list.
If you export the temporary HELPER autofill points and then re-import them, they become fixed (white) shaping points.
The distance HELPER markers can usefully be exported for external use but for Cruiser itself it does not make much sense.
In their temporary state, you can very easily create and manipulate them with undo redo or quickly even produce new ones.

**Reply by devemux86** [reply]
We can change the title to something like "Auto add shaping points".

Changing color, etc. is more complicated, as we'll have to indicate their role in the storage formats.
i.e. change all storage formats

However, these are not really different from the usual shaping points.
Via points only participate in navigation instructions, all shaping points have an auxiliary role.

**Reply by 0709wiwiwi** [reply]
> indicate their role in the storage formats.

I see them most useful in their temporary state. 
Thus, you don't usually need to store them either.
Advice is don't export them and if you still do, export simply as normal (white) shaping points to be usefully used for external gps like the Garmin Zumos. @WolfgangZ42

**Reply by 0709wiwiwi** [reply]
> Their purpose, when introduced, was actually the export to third-party devices so they can follow closely the original route without too many changes.

Correct.

> If we indicate them differently and not store them in storage formats then a restart of Cruiser will return them to their normal shaping state.

Emux & all other.  I need to think about it further practically. 
Reference is the existing [kurviger design](https://github.com/devemux86/cruiser/discussions/497#discussioncomment-9116255) with only start and end. 
Surely the purpose of those temporary extra shaping markers is @ the navigation to prevent an autorecalculation massacre for your entire route ?  For Cruiser use, you don't need to save those temporary helper markers in a file export anyway ?  In a pinch you can re-produce them in just about any distance variation you want.  Just pausing now and waiting for the other(s) opinions first.

**Reply by devemux86** [reply]
> For Cruiser use, you don't need to save those temporary helper markers anyway ?

There is need to not lose them:
we navigate, make a stop, close Cruiser, restart Cruiser,
these extra shaping points should be in their place to continue navigation.

Users are not expected to create them constantly.

**Reply by 0709wiwiwi** [reply]
Roger.

**Reply by devemux86** [reply]
This tool can also be useful in route planning in addition to navigation.

For example, automatically add extra shaping points and then change the section profiles.
So it is not yet certain that they should differ from the "normal" shaping points.

We will probably see it in practice. :slightly_smiling_face:

**Reply by 0709wiwiwi** [reply]
I didn't want to get ahead and first test what's coming first, but you're are already triggering a reaction anyway.
If you import a naked track now, the current tool allows you to convert it by automatic shaping points addition.
The current tool distributes the extra (helper) shaping points quite irregularly sometimes with a lot of density and sometimes too little.
So I am wondering whether it could not be done more optimally by the method with the more regularly distributed distance points ?

**Reply by devemux86** [reply]
Add extra points with similar distance on an existing route is more simple.

When we import a track, the route is calculated after we set the points, so distances may be not similar.
These points' purpose is not the distance, but calculate a route close to the track.

**Comment by devemux86** [comment]
With label the distance to destination (constant in rerouting):

<img src="https://github.com/devemux86/cruiser/assets/3484020/befdc0b8-5d94-4d7a-a36a-4f1a85fe6790" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/de2f6633-4a4a-411a-b624-893757455959" width="50%"/>

**Reply by bios4** [reply]
The "arrow down" is meant for "distance to target"?
Or does it represent a "lower-prio auto-fill-added shaping point"?

And does your right screenshot ^^ tell that "shaping point 29 was available already BEFORE the auto-fill shaping points were added"? Because it has a different description...

**Reply by devemux86** [reply]
> The "arrow down" is meant for "distance to target"?

Correctly, like in the via points.

> And does your right screenshot ^^ tell that "shaping point 29 was available already BEFORE the auto-fill shaping points were added"?

"Shaping point 29" was available before and shows an automatic label, because it has no name.

**Comment by devemux86** [comment]
This [report](https://github.com/devemux86/cruiser/discussions/482#discussioncomment-9172726) will be fixed in the next update.
Distance 1 km can return some extraordinary results. :slightly_smiling_face: 

<img src="https://github.com/devemux86/cruiser/assets/3484020/d14b9584-e1c0-4e70-9536-f1435cca81d4" width="50%"/>

**Comment by bios4** [comment]
Following observation (Cruiser 3.0.43.b15):
- I set up the section profile colors as shown in screenshot 1
- A route with different section profiles is loaded, which are indicated by the selected colors; screenshot 2
- After selecting "Fill with shaping points", these colors are "overwritten" by the default route color where the new shaping points are added; screenshot 3

<img src="https://github.com/devemux86/cruiser/assets/145460147/0b14aca0-33f0-4fcd-8101-e0922ae93016" width="32%">
<img src="https://github.com/devemux86/cruiser/assets/145460147/faf38f55-ccf7-4eb8-9e1b-a99c9848af8e" width="32%">
<img src="https://github.com/devemux86/cruiser/assets/145460147/71a59e33-7cdf-4f85-b16b-b80d6eb91f61" width="32%">

Questions to @devemux86:
- Is this a desired behaviour?
- If yes: would it be possible to preserve the section profile colors when filling with shaping points?

**Reply by devemux86** [reply]
@bios4 indeed this should not happen, I will fix it.

Thanks for the report! :+1:

**Comment by devemux86** [comment, marked as answer]
Fill route with shaping points with adjustable distance is available in [Cruiser 4.0.0](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by WolfgangZ42** [comment]
Tried it today and got this 
<img width="200" src="https://github.com/devemux86/cruiser/assets/121806238/7cf61dff-0ade-42a8-a5d3-4c0b67858120"/>
I expected the distance to destination as mentioned in this thread earlier. 
And I wonder why the first three waypoints are specially marked.
My kurviger file: [Regensburg_2.kurviger.txt](https://github.com/devemux86/cruiser/files/15212012/Regensburg_2.kurviger.txt)

**Reply by 0709wiwiwi** [reply]
I see this.  By default profile = weighting:curvature
Except the first track section toward the shaping point 1 is by profile Fastest. 
Existing shaping points (kurviger origin) are shown as Shaping point  xx or name
Extra shaping points by "fill in" are shown as Shaping point xx or (if set)  - km  distance to target.
![image](https://github.com/devemux86/cruiser/assets/54167201/7e2d691e-0c4d-4003-98d9-23254b69dcd1)

**Reply by devemux86** [reply]
@WolfgangZ42 

The route file is the original route and your screenshot is the result?

The route contains only 1 section profile "fastest" at shaping point 1.
(like in the screenshot by @0709wiwiwi)

The added shaping points get the profile of the route section that includes them.

To have name the distance to destination, please check "Add name" in the dialog.

**Reply by WolfgangZ42** [reply]
I'm embarrassed! I forgot that I set "fastest" to leave the city straight away. 🥴
and overlooked "Add name". 🙄 Sorry!

**Reply by 0709wiwiwi** [reply]
Anyway, perfect for practice. With interesting and popular break spots, by the way.

**Comment by Cer-o** [comment]
Hi everyone, I'm excited about the progression of the app, although I don't really use most of the new features.

This one, for example, interests me a lot and I would like to know how I find it to activate it.

Thank you very much for the great work you are doing.

**Reply by bios4** [reply]
> how I find it to activate it

Long-press on a loaded route > 2nd menu screen ("More ...") > "Fill with shaping points"

**Comment by Cer-o** [comment]
It is a super interesting novelty and it is going to help us a lot.
Thank you so much.
